js/src/components/paid-ads/convertToAssetGroupUpdateBody.test.js

```diff
@@ -0,0 +1,450 @@
+/**
+ * External dependencies
+ */
+import { uniqueId } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import convertToAssetGroupUpdateBody, {
+	diffAssetOperations,
+} from './convertToAssetGroupUpdateBody';
+import { ASSET_KEY, ASSET_GROUP_KEY, ASSET_FORM_KEY } from '.~/constants';
+import { structuredClone } from '.~/utils/structuredClone.js';
+
+function genId() {
+	return Number( uniqueId() );
+}
+
+describe( 'convertToAssetGroupUpdateBody', () => {
+	let assetGroup;
+	let values;
+
+	beforeEach( () => {
+		assetGroup = {
+			[ ASSET_GROUP_KEY.FINAL_URL ]: 'https://example/cap',
+			[ ASSET_GROUP_KEY.DISPLAY_URL_PATH ]: [ 'cap', 'specials' ],
+			assets: {
+				[ ASSET_KEY.BUSINESS_NAME ]: {
+					id: genId(),
+					content: 'General Clothing',
+				},
+				[ ASSET_KEY.HEADLINE ]: [
+					{
+						id: genId(),
+						content: 'headline 1',
+					},
+				],
+				[ ASSET_KEY.LOGO ]: [
+					{
+						id: genId(),
+						content: 'https://example/logo-1.png',
+					},
+				],
+			},
+		};
+
+		values = {
+			[ ASSET_FORM_KEY.FINAL_URL ]: 'https://example',
+			[ ASSET_FORM_KEY.DISPLAY_URL_PATH ]: [ 'cap', 'specials' ],
+			[ ASSET_FORM_KEY.BUSINESS_NAME ]: 'General Clothing',
+			[ ASSET_FORM_KEY.HEADLINE ]: [ 'headline 1' ],
+			[ ASSET_FORM_KEY.LOGO ]: [ 'https://example/logo-1.png' ],
+		};
+	} );
+
+	describe( 'should base on an existing asset entity group and convert the assets form values into the request body for updating that existing asset group', () => {
+		test( 'the returned data should meet the required structure', () => {
+			values[ ASSET_FORM_KEY.FINAL_URL ] = 'https://example/belt';
+			values[ ASSET_FORM_KEY.DISPLAY_URL_PATH ] = [ 'belt', 'specials' ];
+
+			const body = convertToAssetGroupUpdateBody( assetGroup, values );
+
+			expect( body ).toHaveProperty( 'final_url' );
+			expect( body ).toHaveProperty( 'path1' );
+			expect( body ).toHaveProperty( 'path2' );
+			expect( body ).toHaveProperty( 'assets' );
+		} );
```

Comment:
I think we are covering this with the further next 2.

Reply:
The target of this test is to examine whether the returned body has the required properties regardless of the properties' value; The targets of the next two are to examine whether the value of properties matched the expectations.

Although they appear to overlap, when the tests fail, the cause can be located more quickly. For example, if the `final_url` is miswritten to `finalUrl`,
- without this test, the failure will be the following, and whether it's caused by mismatched value or the incorrect property will need further confirming.
   ![image](https://user-images.githubusercontent.com/17420811/219282579-edd64ea7-11fc-4b23-ad77-7c150ea22589.png)
- with this test, the failure will plus the following, and the cause is clearly pointed out that the property name is mismatched.
   ![image](https://user-images.githubusercontent.com/17420811/219282685-1a0ea0c7-7e3a-47d8-a50c-8d610acbe32e.png)